include/envoy/filesystem/filesystem.h

```diff
@@ -1,24 +1,106 @@
 #pragma once
 
-#include <cstdint>
-#include <functional>
 #include <memory>
 #include <string>
 
+#include "envoy/common/platform.h"
 #include "envoy/common/pure.h"
+#include "envoy/event/dispatcher.h"
+#include "envoy/thread/thread.h"
 
 #include "absl/strings/string_view.h"
 
 namespace Envoy {
 namespace Filesystem {
 
 /**
- * Abstraction for a file on disk.
+ * Abstraction for a basic file on disk.
  */
 class File {
 public:
   virtual ~File() {}
 
+  /**
+   * Open the file with O_RDWR | O_APPEND | O_CREAT
+   * The file will be closed when this object is destructed
+   */
+  virtual void open() PURE;
```

Comment:
Right now, open throws if it fails. I think right now the semantics of this `RawFile` class are a bit odd:
- constuctor calls open(), though "closed" is a valid state
- open() throws on failure
- write() returns an error code on failure
- close() swallows any errors

I think it might make more sense for:
- constructor doesn't do anything
- open/write/close return the error codes

And then the user of this  class (the file object with stats+flushing) can handle the return codes / throw exceptions as before

Reply:
+100 but I also appreciate you don't want to change semantics in a giant PR.

Can you leave a TODO?

Reply:
No, I think we should get the semantics correct here -- this is a new object. Since it is really only used by the File object that has stats + flushing at the moment, the important thing to is to maintain those semantics, which should be straightforward with the proposed `RawFile` semantics